src/Lifecycle/StateManager.php

```diff
@@ -188,18 +200,47 @@ protected function loadMany(iterable $ids, string $type): StateCollection
     protected function reconstitute(State $state): static
     {
         // When we're replaying, the Broker is in charge of applying the correct events
-        // to the State, so we only need to do it *outside* of replays.
-        if (! $this->is_replaying) {
-            $this->events
-                ->read(state: $state, after_id: $state->last_event_id)
-                ->each(fn (Event $event) => $this->dispatcher->apply($event));
-
-            // It's possible for an event to mutate state out of order when reconstituting,
-            // so as a precaution, we'll clear all other states from the store and reload
-            // them from snapshots as needed in the rest of the request.
-            // FIXME: We still need to figure this out
-            // $this->states->reset();
-            // $this->remember($state);
+        // to the State, so we need to skip during replays. Similarly, if we're already
+        // reconstituting in a recursive call, the root call is responsible for applying
+        // events, so we should also skip in that case.
+
+        if (! $this->is_replaying && ! $this->is_reconstituting) {
+            try {
+                $this->is_reconstituting = true;
+
+                $summary = $this->events->summarize($state);
+
+                // FIXME: We probably need to re-write all the snapshots after we're done
+                // FIXME: Swap out existing state manager, push all related states into new state manager
+                // FIXME: run all the event on them, swap them out
+
+                $this->events->get($summary->related_event_ids)
+                    ->filter(function (Event $event) {
+                        $last_event_ids = $event->states()
+                            ->map(fn (State $state) => $state->last_event_id)
+                            ->filter();
+
+                        $min = $last_event_ids->min() ?? PHP_INT_MIN;
+                        $max = $last_event_ids->max() ?? PHP_INT_MIN;
+
+                        // If all states have had this or future events applied, just ignore them
+                        if ($min >= $event->id && $max >= $event->id) {
+                            return false;
+                        }
+
+                        // We should never be in a situation where some events are ahead and
+                        // others are behind, so if that's the case we'll throw an exception
+                        if ($max > $event->id && $min <= $event->id) {
+                            throw new RuntimeException('Trying to apply an event to states that are out of sync.');
+                        }
```

Comment:
This is all to account for cases where one state is ahead of another. We may want to decide that in the beginning we just don't support that…?